### ISSUE TITLE
Using Neovide on macOS

### ISSUE THREAD

**Issue by casperstorm** [Kethku/neovide#301]
Up until now I have been building the binary my self with `cargo run --release` and then used that binary with `neovide . &` – using the `&` so my terminal doesn't wait for `neovide` to return. This works. 

However, then don't have the `Neovide` icon etc., so I thought I wanted to try the prebuilt `.app` on the [release](https://github.com/Kethku/neovide/releases) page. But that app just instantly closes, and can't start.

I have tried to look at https://github.com/Kethku/neovide/issues/46 and https://github.com/Kethku/neovide/issues/167 but they all seem to have problems when building it – which I have not. I just want to be able to run the pre-built `.app`. Any ideas anyone?

Running 10.15.5.

**Comment by clason** [comment]
The executable for the actual binary in the package isn't set -- if you `cd` into `Neovide.app` from the terminal down until you find `neovide` and then `chmod +x neovide`, it works (after the usual security shenanigans on current macOS).

**Comment by casperstorm** [comment]
> The executable for the actual binary in the package isn't set -- if you `cd` into `Neovide.app` from the terminal down until you find `neovide` and then `chmod +x neovide`, it works (after the usual security shenanigans on current macOS).

I forgot to write that I've also tried this. No luck though.
Nothing to approve in Security & Privacy as some has mentioned in another issue as well.

**Comment by j4qfrost** [comment]
I am currently investigating. We had solved a similar issue before, so I'll roll back to see where something might have broken.

**Comment by clason** [comment]
You've got to try a few times, apparently, if the app isn't notarized. Have you tried launching from the command line (`./neovide` in the directory mentioned above)? That worked for me.

**Comment by j4qfrost** [comment]
I can use the terminal just fine and ```open -a <full path>/Neovide.app``` works for me, but this behavior indicates a possible pathing issue.

**Comment by casperstorm** [comment]
> Oh, wow, that doesn't work anymore, either. It's the same binary as before, but it looks like macOS put on additional tin foil hats for 10.15.5 (which just got released, so that's the only difference between the two attempts).

I would like to point out that I tried on 10.15.3 yesterday, had the same issue and then just updated to 10.15.5 this morning.

**Comment by clason** [comment]
What's your macOS version? I just get `Unable to find application named 'Downloads/Neovide.app'`.

**Comment by clason** [comment]
OK, now it works. I had to right-click to open multiple times and select "yes, really" (after setting the executable bit) for it to work.

So your issue is different from mine -- what's your machine (i.e., what GPU are you using)? It may be a driver issue.

**Comment by casperstorm** [comment]
> What's your macOS version? I just get `Unable to find application named 'Downloads/Neovide.app'`.

I've tested on both `10.15.3` and `10.15.5` – both doesn't work for me.

> So your issue is different from mine -- what's your machine (i.e., what GPU are you using)? It may be a driver issue.

I have a Macbook Pro 16". AMD Radeon Pro 5500M dedicated GPU.

But since I can build the binary my self from `cargo run --release` it smells of a issue when you guys are building the releases, no?

**Comment by j4qfrost** [comment]
There's some weird pathing behavior in OSX that causes this issue to happen for double clicking i.e. opening from the launcher as opposed to running neovide from terminal. The behavior is similar to this https://github.com/Kethku/neovide/issues/202

**Comment by clason** [comment]
Yes, and it's not specific to Neovide (also happens for Goneovim, for example).

(Neovide also has problems with non-standard shells (fish), where the providers aren't picked up correctly (at all), but that's a different issue -- possibly related to pathing, which is why I mention it here.)

**Comment by j4qfrost** [comment]
I'll write up a fix now, I narrowed it down to a commit that I pushed. I'll ping you guys in a bit.

**Comment by casperstorm** [comment]
I just have a small follow up. I built it my self doing what you guys are doing in the github action:

```
    cargo build --release
    cargo bundle --release
    cd target/release/bundle/osx/Neovide.app/Contents
    mkdir Frameworks
    cp /usr/local/lib/libvulkan.dylib ./Frameworks/
    cp /usr/local/lib/libMoltenVK.dylib ./Frameworks/
    mkdir -p Resources/vulkan
    cp -r /usr/local/share/vulkan/icd.d ./Resources/vulkan/
    jq '.ICD.library_path = "../../../Frameworks/libMoltenVK.dylib"' 
    ./Resources/vulkan/icd.d/MoltenVK_icd.json > MoltenVK_icd.json
    mv MoltenVK_icd.json ./Resources/vulkan/icd.d/
    install_name_tool -add_rpath "@executable_path/../Frameworks" ./MacOS/neovide
```
This gave me a bundled `.app` which I can run.

**Comment by casperstorm** [comment]
EDIT:
Okay guys, I have some more information here. This particular issue has nothing to do with the specific MacOS version, but something to do with the way you are linking dylib's.

The issue is that if you have installed vulkan frameworks yourself, and running the `Neovide.app` it will fail, because it does not know which vulkan dylib to choose because now you have two version of the vulkan dylib. Hence this message:

```
Class MVKBlockObserver is implemented in both  
/Users/crs/Downloads/Neovide.app/Contents/Frameworks/libMoltenVK.dylib (0x112bd9be0) and    
/usr/local/Caskroom/vulkan-sdk/1.2.131.1/macOS/lib/libMoltenVK.dylib (0x11324fbe0).

One of the two will be used. Which one is undefined.
```
This means that running the app on a machine, without vulcan frameworks, will actually work because then it will just use the dylibs you are linking to.

To get an overview of the dylibs on the binary you can run `otool -l ./neovide` by that output you can see:

```
cmd LC_LOAD_DYLINKER                                                                                                                                                                                
      cmdsize 32                                                                                                                                                                                              
         name /usr/lib/dyld (offset 12)                                                                                                                                                                                          
```
and
```
cmd LC_RPATH
      cmdsize 48
         path @executable_path/../Frameworks (offset 12)
```
I believe this is the problem: you are bundling dynamic libraries wrong. I don't have a fix up my sleeve right this moment, because frankly I am not an expert in this as well but I'll look more into it later.

**Comment by j4qfrost** [comment]
```cargo bundle``` actually handles copying the necessary libraries into the .app; neovide works just before the bump from 0.5, so it is most likely not an issue with the library for you guys, and definitely not for me.

I just implemented a fix and can safely say that it isn't a library issue. As I suspected, my commit here https://github.com/Kethku/neovide/commit/b7b4e95ca97cc29933dd70f34c4274229a9b0d95 was the cause. Macos can't locate nvim correctly through the launcher so as a quick fix, I had the command function fallback to the hardcoded value. I'm not an OSX developer, so I'm not confident about what the best fix is.

I'll push the patch in a bit. I had some trouble writing a more elegant solution, but it's finished.

**Comment by j4qfrost** [comment]
https://github.com/Kethku/neovide/pull/302 @casperstorm @clason if you guys can confirm that this works, I'll merge it in. Run ```cargo bundle --release``` then ```chmod +x``` the binary inside.

**Comment by casperstorm** [comment]
I am not quite sure what you are trying to achieve with that patch, @j4qfrost and also hardcoding the path to `nvim` does not seem like a proper solution. What if I have it somewhere else?

```rust
let default_path = "/usr/local/bin/nvim";
```
It should properly be a preference to select the path to it, at least.

> cargo bundle actually handles copying the necessary libraries into the .app; neovide works just before the bump from 0.5, so it is most likely not an issue with the library for you guys, and definitely not for me.

I don't know what problem you are facing, but I am differently facing a libraries issue. You guys are doing a lot of strange stuff AFTER `cargo bundle` and that is where I believe it is going wrong:

```sh
cp /usr/local/lib/libvulkan.dylib ./Frameworks/
cp /usr/local/lib/libMoltenVK.dylib ./Frameworks/
[...]
cp -r /usr/local/share/vulkan/icd.d ./Resources/vulkan/
jq '.ICD.library_path = "../../../Frameworks/libMoltenVK.dylib"' 
[...]
install_name_tool -add_rpath "@executable_path/../Frameworks" ./MacOS/neovide
```
As I said, this is an edge case, but if you have installed Vulcan frameworks yourself then the app will not start. That is what I am facing. I can build the `.app` myself and not link the Vulcan frameworks [like you guys are doing](https://github.com/Kethku/neovide/blob/master/.github/workflows/build.yml#L69) because then it will just simply use the once I have (from `brew`), but that is not the case for everyone.

TLDR:
* Your fix is not fixing this problem. It seems you are trying to guard against the case where someone would run it without having `nvim` installed.
* There is a problem with `Neovide` and the way you are doing dylib linking. If you have installed Vulcan from eg. `brew` and then run the executables you provide from the release page, it will fail because it does not know which to choose.

**Comment by j4qfrost** [comment]
This may very well be two separate issues, which present with the same faulty behaviors. I'm sorry I can't really say much else about the linking issue aside from "it works on my machine". I have Vulkan installed through brew.

Sorry, I actually overlooked your error the twice reading your comments:
```Class MVKBlockObserver is implemented in both  
/Users/crs/Downloads/Neovide.app/Contents/Frameworks/libMoltenVK.dylib (0x112bd9be0) and    
/usr/local/Caskroom/vulkan-sdk/1.2.131.1/macOS/lib/libMoltenVK.dylib (0x11324fbe0).

One of the two will be used. Which one is undefined.
```
I saw a problem as a nail and try to hammer it in. Anyway can you run something like this and see if it works:

```install_name_tool -change existing_path_to_dylib @executable_path/blah.dylib executable_name```

It looks like there's still work being done on this https://github.com/KhronosGroup/Vulkan-Loader/issues/300

Side note:
To solve for this issue that you have nvim installed somewhere else There is a NEOVIM_BIN variable you can set that overrides the default. Previously we used which to figure out the first available nvim binary to use, but it seemed to be failing on OSX. I built the previous commits and they worked and my fork also works.

**Comment by j4qfrost** [comment]
https://github.com/KhronosGroup/Vulkan-Loader/blob/master/loader/LoaderAndLayerInterface.md#icd-discovery-on-macos @casperstorm I'm curious about which of the locations you have on your machine. I have the manifest in /usr/local/share/vulkan/icd.d

**Comment by casperstorm** [comment]
> https://github.com/KhronosGroup/Vulkan-Loader/blob/master/loader/LoaderAndLayerInterface.md#icd-discovery-on-macos @casperstorm I'm curious about which of the locations you have on your machine. I have the manifest in /usr/local/share/vulkan/icd.d

```
$ ll /usr/local/share/vulkan/     
total 0
drwxr-xr-x  5 crs  admin   160B Feb 13 11:22 explicit_layer.d
drwxr-xr-x  3 crs  admin    96B Feb 13 11:22 icd.d
```

**Comment by j4qfrost** [comment]
What's inside your /usr/local/share/vulkan/icd.d/MoltenVK_icd.json  file?

**Comment by casperstorm** [comment]
```
$ bat  /usr/local/share/vulkan/icd.d/MoltenVK_icd.json       
   
   1   │ {
   2   │   "file_format_version": "1.0.0",
   3   │   "ICD": {
   4   │     "library_path": "/usr/local/lib/libMoltenVK.dylib",
   5   │     "api_version": "1.0.0"
   6   │   }
   7   │ }

```

**Comment by j4qfrost** [comment]
Interesting, I have this line
```"library_path": "../../../lib/libMoltenVK.dylib", ```

**Comment by casperstorm** [comment]
> Interesting, I have this line
> `"library_path": "../../../lib/libMoltenVK.dylib", `
> 
> When did you install Vulkan?

Hmm, not long again, a few weeks.
Have you tried to reinstall?

**Comment by j4qfrost** [comment]
I just reran brew cask install apenngrace/vulkan/vulkan-sdk and the configuration file didn't change. Though neovide works just fine for me. No clue how our vulkan loader manifests diverged.

**Comment by ishammahajan** [comment]
I don't know if this will help, but this issue is strange because running `open ~/Downloads/Neovide.app` does open the app, but double clicking doesn't. I have replaced the binary inside with the one I built from the terminal, and removed the libraries present in `Contents/Frameworks` folder.

**Comment by j4qfrost** [comment]
@ishammahajan try this branch https://github.com/Kethku/neovide/pull/302

I suspect that there are two separate problems being addressed in this thread. We should probably split this ticket soon.

**Comment by ishammahajan** [comment]
Thanks, that branch works! I'm a little curious though, how do the fixes in that branch make things work? The issue with having a full path to `nvim` was resolved earlier, right?

**Comment by j4qfrost** [comment]
@ishammahajan I made the changes as an enhancement request from another user that would search through the PATH first in order to find where nvim was installed. The issue is that when launching from double-click PATH isn't set, but I overlooked that problem and didn't add a check in the case that no binary was found. In the branch that I linked, I add the check.

@casperstorm Have tried editing the library path in ```/usr/local/share/vulkan/icd.d/MoltenVK_icd.json``` to see if neovide runs?

**Comment by ishammahajan** [comment]
Thanks @j4qfrost for the answer and all your work on this :)

**Comment by bsander** [comment]
Not sure how related this is, but with #302 I can double-click the Neovide app now and get it working, but coc.nvim starts complaining about not being able to find node, and indeed the `$PATH` variable does not contain any of my regular path extensions (like `/usr/local/bin` for homebrew). Is there any way to deal with that?

**Comment by clason** [comment]
I suspect neovide doesn't properly pickup on the default (login) shell settings. I see the following behavior:

1. opening `./neovide` inside the app from the terminal, everything works fine;
2. double-clicking the app in Finder and running `:checkhealth`, I get 
```
health#provider#check
========================================================================
  - ERROR: Failed to run healthcheck for "provider" plugin. Exception:
    function health#check[21]..health#provider#check[5]..<SNR>80_check_ruby[13]..<SNR>80_system, line 11
    Vim(let):E903: Process failed to start: no such file or directory: "fish"
```
(not sure where `fish` is picked up from -- I set this as shell in the terminal, not as login shell through Preferences).

**Comment by j4qfrost** [comment]
@clason you're correct, here's an explanation https://stackoverflow.com/questions/8918685/mac-app-crashes-from-finder-but-runs-inside-terminal

@bsander Can you open up another issue for this? I know how to fix the problem, but this is starting to lose focus a bit, because the issue was a bit generic to begin with.

@casperstorm Also it would be helpful if you change open up another issue to specify that your problem is specific to Vulkan not loading correctly.

**Comment by casperstorm** [comment]
> @casperstorm Also it would be helpful if you open up another issue to specify that your problem is specific to Vulkan not loading correctly.

I'll make a note, dig into it and re-open one. Feel free to close this when ready. Sorry for long delay in response.
